### PR TITLE
Get x_work when changing m_state

### LINKED ISSUE #4552: A deadlock was observed in "kademlia" test
https://travis-ci.org/ethereum/cpp-ethereum/jobs/279673618 timeouts with

```
Test Case "kademlia": 
.
.
.
.
.
```

### PR DESCRIPTION
**Pull request by pirapira**
Fixes #4552 

This fixes the deadlock in "kademlia" test.  The problem happned when the worker thread runs so quickly that when the main thread starts waiting for `m_state` change, the thread has already finished working.  This resulted in the main thread waiting forever for `m_state` to become `Started`.

### CODE REVIEW

**Comment by pirapira** [comment]
I was confused about assignees and reviewers on this PR, it seems.

**Comment by gumb0** [comment]
I've got the notification anyway

File: libdevcore/Worker.cpp [hunk 1]
```diff
@@ -48,9 +48,13 @@ void Worker::startWorking()
 			while (m_state != WorkerState::Killing)
 			{
 				WorkerState ex = WorkerState::Starting;
-				bool ok = m_state.compare_exchange_strong(ex, WorkerState::Started);
+				{
+					// When somebody waits for this change on m_state, they'll give me the lock saying 'm_state_notifier.wait(l)'
+					Guard l(x_work);
```
Comment (gumb0) [inline libdevcore/Worker.cpp]: Let's use `std::unique_lock`, because it's used in other places of this class

File: libdevcore/Worker.cpp [hunk 2]
```diff
@@ -48,9 +48,13 @@ void Worker::startWorking()
...
+					bool ok = m_state.compare_exchange_strong(ex, WorkerState::Started);
```
Comment (gumb0) [inline libdevcore/Worker.cpp]: I think we can just ignore the returned value instead of `(void)ok;`, no?

File: libdevcore/Worker.cpp [hunk 3]
```diff
@@ -48,9 +48,13 @@ void Worker::startWorking()
 			while (m_state != WorkerState::Killing)
 			{
 				WorkerState ex = WorkerState::Starting;
-				bool ok = m_state.compare_exchange_strong(ex, WorkerState::Started);
+				{
+					// When somebody waits for this change on m_state, they'll give me the lock saying 'm_state_notifier.wait(l)'
```
Comment (gumb0) [inline libdevcore/Worker.cpp]: I'd just say that it's condition_variable-related mutex, maybe it's clearer...

File: libdevcore/Worker.cpp (hunk 2 above)
Comment (pirapira) [reply libdevcore/Worker.cpp]: Should be the same thing.  Probably it's just fine to `m_state = Started;`.

File: libdevcore/Worker.cpp (hunk 3 above)
Comment (pirapira) [reply libdevcore/Worker.cpp]: OK.

**Comment by pirapira** [comment]
I rebased and took @gumb0's suggestions.

**Pull request merged**